pkgs/misc/crust/default.nix

```diff
@@ -0,0 +1,154 @@
+###!
+###! When packaging for linux make sure that you have the following patches otherwise crust won't be able to cleanly share the clock controller and PMIC bus controller hardware: https://github.com/torvalds/linux/compare/master...crust-firmware:linux:crust
+###!
+###! If you are faced with issues while building the firmware then take a look at https://github.com/crust-firmware/crust#building-the-firmware
+
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchpatch
+, bison
+, flex
+, buildPackages
+}:
+
+###! The package declaration is heavily inspired by the U-Boot package to make maintenance easier, but should stay as an invidual declaration since it's processing has different requirements and including it in the U-Boot would bloat the definition and would be harder to understand and maintain
+
+let
+  defaultVersion = "unstable-2023-03-04";
+  defaultSrc = fetchFromGitHub {
+    owner = "crust-firmware";
+    repo = "crust";
+    rev = "c308a504853e7fdb47169796c9a832796410ece8";
+    sha256 = "AobVD3Jo/6s0cExHXpYYAqZv/gCplxLlDYVscSCIS6M=";
+  };
+  buildCrust = {
+    version ? "${defaultVersion}"
+  , src ? null
+  , filesToInstall
+  , installDir ? "$out"
+  , defconfig
+  , extraConfig ? ""
+  , extraPatches ? []
+  , extraMakeFlags ? []
+  , extraMeta ? {}
+  , ... } @ args: stdenv.mkDerivation ({
+    # Strip "_defconfig" suffix from 'defconfig' if it's present for clearer output
+    # builtins.match returns 'Null' if the string matches so we have to then check if it's Null
+    # FIXME-QA(Krey): I hate this, but am not aware of better way to write this
+    pname = if (builtins.isNull ((builtins.match "_defconfig$" defconfig)))
+      then "crust-${lib.strings.removeSuffix "_defconfig" defconfig}"
+      else "crust-${defconfig}";
+
+    ###! The package has upstream definitions of configurations for set devices and thus the defconfig has to be always included, beyond that the options are designed to be as flexible as possible to accompany special requirements of various devices
+    ###!
+    ###! Avoid declaring extra configuration (extraConfig) when possible as those should go to the upstream unless you use special usecase device e.g. modded teres-A64
+
+    defconfig = if defconfig == null then
+      throw "crust config not defined!"
+      else defconfig;
+
+    version = if src == null then
+      defaultVersion
+      else version;
+
+    src = if src == null then
+      defaultSrc
+      else src;
+
+    # Only for global crust patches, per-device paches should be included as extras to avoid adding noise which would complicate debugging
+    patches = [] ++ extraPatches;
+
+    postPatch = "";
+
+    nativeBuildInputs = [
+      flex
+      bison
+    ];
+
+    depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+    makeFlags = [
+      # This is double-quotted to prevent string splitting (SC2086)
+      "CROSS_COMPILE=\"${stdenv.cc.targetPrefix}\""
+    ] ++ extraMakeFlags;
+
+    passAsFile = [ "extraConfig" ];
+
+    ###! Phases
```

Comment:
```suggestion
```

Reply:
This is a specification, it's there to ensure that people are not declaring phases above the code for readability